In [1]:
import pandas as pd
import statistics
import numpy as np

### Q1. Pandas version

What's the version of Pandas that you installed?

In [2]:
pd.__version__

'2.3.1'

In [3]:
df = pd.read_csv("https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv")

### Q2. Records count

How many records are in the dataset?

In [4]:
df

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                     170            3.0       159.0     3413.433759   
1                     130            5.0        97.0     3149.664934   
2                     170            NaN        78.0     3079.038997   
3                     220            4.0         NaN     2542.392402   
4                     210            1.0       140.0     3460.870990   
...                   ...            ...         ...             ...   
9699                  140            5.0       164.0     2981.107371   
9700                  180            NaN       154.0     2439.525729   
9701                  220            2.0       138.0     2583.471318   
9702                  230            4.0       177.0     2905.527390   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year  origin fuel_type         drivetrain  \
0             17.7        2003  Europe  Gasoline    All-wheel drive   
1             17.8        2007     USA  Gasoline  Front-wheel drive   
2             15.1        2018  Europe  Gasoline  Front-wheel drive   
3             20.2        2009     USA    Diesel    All-wheel drive   
4             14.4        2009  Europe  Gasoline    All-wheel drive   
...            ...         ...     ...       ...                ...   
9699          17.3        2013  Europe    Diesel  Front-wheel drive   
9700          15.0        2004     USA  Gasoline    All-wheel drive   
9701          15.1        2008     USA    Diesel    All-wheel drive   
9702          19.4        2011     USA    Diesel  Front-wheel drive   
9703          14.7        2005    Asia    Diesel    All-wheel drive   

      num_doors  fuel_efficiency_mpg  
0           0.0            13.231729  
1           0.0            13.688217  
2           0.0            14.246341  
3           2.0            16.912736  
4           2.0            12.488369  
...         ...                  ...  
9699        NaN            15.101802  
9700        0.0            17.962326  
9701       -1.0            17.186587  
9702        1.0            15.331551  
9703       -1.0            14.884467  

[9704 rows x 11 columns]

A2: 9704 records

### Q3. Fuel types

How many fuel types are presented in the dataset?

In [5]:
df["fuel_type"].nunique()

2

A3: 2 fuel types

### Q4. Missing values

How many columns in the dataset have missing values?

In [6]:
df.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

A4: 4 columns have missing values

### Q5. Max fuel efficiency

What's the maximum fuel efficiency of cars from Asia?

In [7]:
df.groupby("origin").fuel_efficiency_mpg.max()

origin
Asia      23.759123
Europe    25.967222
USA       24.971452
Name: fuel_efficiency_mpg, dtype: float64

A5: 23.75 (round down)

### Q6. Median value of horsepower


1. Find the median value of `horsepower` column in the dataset.
2. Next, calculate the most frequent value of the same `horsepower` column.
3. Use `fillna` method to fill the missing values in `horsepower` column with the most frequent value from the previous step.
4. Now, calculate the median value of `horsepower` once again.

Has it changed?

In [8]:
statistics.median(df["horsepower"])

165.0

In [9]:
mode = statistics.mode(df["horsepower"])
print(mode)

152.0


In [10]:
df["horsepower"]=df["horsepower"].fillna(value=mode)

In [11]:
statistics.median(df["horsepower"])

152.0

Original median: 165.0 \
Mode (most frequent value): 152.0 \
Median after filling missing values: 152.0

A6: Yes, it decreased.

### Q7. Sum of weights

1. Select all the cars from Asia
2. Select only columns `vehicle_weight` and `model_year`
3. Select the first 7 values
4. Get the underlying NumPy array. Let's call it `X`.
5. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
6. Invert `XTX`.
7. Create an array `y` with values `[1100, 1300, 800, 900, 1000, 1100, 1200]`.
8. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
9. What's the sum of all the elements of the result?

In [12]:
df2=df[(df["origin"]=="Asia")]
df2 = df2[["vehicle_weight", "model_year"]]
first_7=df2.iloc[:7]
first_7

vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016
34     2844.227534        2014
38     3761.994038        2019

In [13]:
X=np.array(first_7)
XT=np.transpose(X)
XTX = XT.dot(X)
inv_XTX = np.linalg.inv(XTX)
y=np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
w=inv_XTX.dot(XT).dot(y)
print(w)
print(sum(w))

[0.01386421 0.5049067 ]
0.5187709081074016


A7: Sum of all elements of the result = 0.51 (round down)